applicationset/controllers/applicationset_controller.go

```diff
@@ -1312,4 +1316,47 @@ func syncApplication(application argov1alpha1.Application, prune bool) (argov1al
 	return application, nil
 }
 
+var ownsHandler = predicate.Funcs{
+	CreateFunc: func(e event.CreateEvent) bool {
+		// if we are the owner and there is a create event, we most likely created it and do not need to
+		// re-reconcile
+		log.Debugln("received create event from owning an application")
+		return false
+	},
+	DeleteFunc: func(e event.DeleteEvent) bool {
+		log.Debugln("received delete event from owning an application")
+		return true
+	},
+	UpdateFunc: func(e event.UpdateEvent) bool {
+		log.Debugln("received update event from owning an application")
+		appOld, isApp := e.ObjectOld.(*argov1alpha1.Application)
+		if !isApp {
+			return false
+		}
+		appNew, isApp := e.ObjectNew.(*argov1alpha1.Application)
+		if !isApp {
+			return false
+		}
+		// don't change the event objects
+		appOld = appOld.DeepCopy()
+		appNew = appNew.DeepCopy()
+		// the applicationset controller only owns Application.Spec and Application.Metadata.
+		// we do not need to re-reconcile if parts of the application changes outside the applicationset's control.
+		// an example being, Application.ApplicationStatus.ReconciledAt which gets updated by the application controller.
+		// Application.ObjectMeta.ResourceVersion and Application.ObjectMeta.Generation are set by K8s.
+		appOld.ObjectMeta.ResourceVersion = ""
+		appOld.ObjectMeta.Generation = 0
+		appNew.ObjectMeta.ResourceVersion = ""
+		appNew.ObjectMeta.Generation = 0
```

Comment:
Shall `ObjectMeta.ManagedFields` also be excluded? There are many fields changing in there (ex: `ObjectMeta.ManagedFields.Time` and `ObjectMeta.ManagedFields.FieldsV1) that are triggering the reconciliation of the whole applicationSet when an app is synched.

This occurs either when synching an app from the UI, or periodically, for example if having two apps missconfigured by owning the same deployment.

```suggestion
		appNew.ObjectMeta.Generation = 0
		appOld.ObjectMeta.ManagedFields = nil
		appNew.ObjectMeta.ManagedFields = nil
```

Reply:
Good shout. I am of the opinion we either ignore`ObjectMeta` completely or explicitly only check annotations/labels. 

Thoughts?

Reply:
I would say explicitly check annotations, labels, and **finalizers**. They are managed by the applicationset and defined here: https://github.com/argoproj/argo-cd/blob/6e02f8b23201b0620a4ff1bce5d38229ba1eb02e/applicationset/controllers/applicationset_controller.go#L591-L594

I also find it much safer than excluding some fields. Redundant reconciles are an expensive operation for the git server, as each reconciliation translates into a `git fetch` operation.

Reply:
Preach, one of the main drivers for this PR!